src/test/kotlin/org/opensearch/replication/repository/RestoreContextTests.kt

```diff
@@ -18,32 +21,54 @@ class RestoreContextTests : OpenSearchTestCase() {
         // given
         val fileName = "test_file"
 
-        val mockShard = mock(IndexShard::class.java)
-        val mockIndexCommit = mock(GatedCloseable::class.java) as GatedCloseable<IndexCommit>
-        val mockStore = mock(Store::class.java)
-        val mockDirectory = mock(org.apache.lucene.store.Directory::class.java)
-        val mockBaseInput = mock(IndexInput::class.java)
-        val mockClonedInput = mock(IndexInput::class.java)
+        val mockShard = mock<IndexShard>()
+        val mockIndexCommit = mock<GatedCloseable<IndexCommit>>()
+        val mockDirectory = mock<Directory>()
+        val mockBaseInput = mock<IndexInput>()
+        val mockClonedInput = mock<IndexInput>()
+        val mockStore = mock<Store> {
+            on { directory() } doReturn mockDirectory
+        }
 
-        // when
-        `when`(mockStore.directory()).thenReturn(mockDirectory)
-        `when`(mockDirectory.openInput(eq(fileName), any(IOContext::class.java))).thenReturn(mockBaseInput)
-        `when`(mockBaseInput.clone()).thenReturn(mockClonedInput)
-        val sut = RestoreContext(
+        whenever(mockDirectory.openInput(eq(fileName), any())).thenReturn(mockBaseInput)
+        whenever(mockBaseInput.clone()).thenReturn(mockClonedInput)
+
+        val sut = object : RestoreContext(
             restoreUUID = UUID.randomUUID().toString(),
             shard = mockShard,
             indexCommitRef = mockIndexCommit,
             metadataSnapshot = Store.MetadataSnapshot.EMPTY,
             replayOperationsFrom = 0L,
-        )
+        ) {
+            override fun withStoreReference(store: Store, block: () -> Unit) {
+                block()
+            }
+        }
+
+        val threads = mutableListOf<Thread>()
+        val results = mutableListOf<IndexInput>()
+        val lock = Object()
+
+        // when
+        repeat(10) {
+            val thread = Thread {
+                val input = sut.openInput(mockStore, fileName)
+                synchronized(lock) {
+                    results.add(input)
+                }
+            }
+            threads.add(thread)
+        }
 
-        val result1 = sut.openInput(mockStore, fileName)
-        val result2 = sut.openInput(mockStore, fileName)
+        threads.forEach { it.start() }
+        threads.forEach { it.join() }
 
         // then
-        assertSame(mockClonedInput, result1)
-        assertSame(mockClonedInput, result2)
-        verify(mockDirectory, times(1)).openInput(eq(fileName), any(IOContext::class.java))
-        verify(mockBaseInput, times(2)).clone()
+        assertEquals(threads.size, results.size)
+        results.forEach {
+            assertSame(mockClonedInput, it)
+        }
+        verify(mockDirectory, times(1)).openInput(eq(fileName), any())
+        verify(mockBaseInput, times(threads.size)).clone()
```

Comment:
i added a test case where multiple threads attempt to open the same file concurrently.
While `clone` is called as many times as the number of threads, the actual `openInput` is invoked only once.